reco_utils/dataset/pandas_df_utils.py

```diff
@@ -74,3 +75,75 @@ def filter_by(df, filter_by_df, filter_by_cols):
             filter_by_df.set_index(filter_by_cols).index
         )
     ]
+
+
+def df_to_libffm(df, col_rating=DEFAULT_RATING_COL, filepath=None):
+    """Converts an input Dataframe (df) to an text file in libffm format.
+
+    Note:
+        The input dataframe is expected to represent the feature data in the following schema
+        |field-1|field-2|...|field-n|rating|
+        |feature-1-1|feature-2-1|...|feature-n-1|1|
+        |feature-1-2|feature-2-2|...|feature-n-2|0|
+        ...
+        |feature-1-i|feature-2-j|...|feature-n-k|0|
+        Where
+        1. each "field-*" occupies one column in the data, and
+        2. "feature-*-*" can be either a string or a numerical value.
+
+        The above data will be converted to the libffm format by following the convention as explained in
+        https://www.csie.ntu.edu.tw/~r01922136/slides/ffm.pdf
+
+    Args:
+        df (pd.DataFrame): input Pandas dataframe.
+        col_rating (str): rating of the data.
+        filepath (str): path to save the converted data.
+
+    Return:
+        df_libffm (pd.DataFrame): a Pandas DataFrame in libffm format.
+    """
+    df_new = df.copy()
+
+    # Encode field (rating column is not included)
+    field_names = list(df_new.drop(col_rating, axis=1).columns)
+    field_dict = {k: v+1 for v, k in enumerate(field_names)}
+
+    # Encode field-feature
+    field_feature_list = []
+    for col in field_names:
+        if df_new[col].dtype == int:
+            df_new[col] = df_new[col].astype(str)
+        if df_new[col].dtype == object:
+            field_feature_concat = ['{}-{}'.format(col, x) for x in df_new[col].values]
+            field_feature_list.extend(field_feature_concat)
+    field_feature_dict = {k: v+1 for v, k in enumerate(field_feature_list)}
```

Comment:
possible to clean this up a bit
```
idx = 1
field_feature_map = {}
for field in field_names:
    if df_new[field].dtype == object:
        for feature in df_new[field].values:
            field_feature_map[(field, feature)] = idx
            idx += 1
````

Reply:
Good point - fixed it based on @gramhagen's codes. 
The older approach will lead to overwriting and thus make the index values not in a sequence.